astropy/stats/bayesian_blocks.py

```diff
@@ -286,24 +277,22 @@ def p0_prior(self, N, Ntot):
         """
         return 4 - np.log(73.53 * self.p0 * (N ** -0.478))
 
-    def gamma_prior(self, N, Ntot):
-        """
-        Basic prior, parametrized by slope gamma.
-        See eq. 3 in Scargle (2012)
-        """
-        if self.gamma == 1:
-            return 0
-        else:
-            return (np.log(1 - self.gamma)
-                    - np.log(1 - self.gamma ** (Ntot + 1))
-                    + N * np.log(self.gamma))
-
     # the fitness_args property will return the list of arguments accepted by
     # the method fitness().  This allows more efficient computation below.
     @property
     def _fitness_args(self):
         return signature(self.fitness).parameters.keys()
 
+    def compute_ncp_prior(self, N):
+        """
+        If ``ncp_prior`` is not explicitly defined, compute it from ``gamma``
+        or ``p0``.
+        """
+        if self.gamma is not None:
+            self.ncp_prior = -np.log(self.gamma)
```

Comment:
Since time is tight right now, what do you think about leaving this as is for now?

I'm thinking of how to let the user specify a prior, and I want to make sure I do it right.

The reason that Scargle chose the geometrical prior is that the math works out so well.
If a user wants an arbitrary function of N_blocks, we could allow them to subclass something like PriorFunc, and define a general p(N_blocks). Then in place of ncp_prior, we could take ln(p(N_blocks+1)) - ln(p(N_blocks)). This would be slow, but if there's a simple closed form of ln(p(N_blocks)), we could allow the user to overload that instead, which would be faster. Just lnp(N_blocks+1) - lnp(N_blocks).

I'm not sure how sound this is. I've fixed all the issues brought up except for this one for now. I'll push what I have for now.